src/jquery.uls.core.js

```diff
@@ -247,9 +235,89 @@
 			} );
 
 			this.$languageFilter.on( 'noresults.uls', $.proxy( lcd.noResults, lcd ) );
+		},
+
+		recreateLanguageFilter: function () {
+			this.$resultsView.removeData( 'lcd' );
+			this.$resultsView.empty();
+			this.$languageFilter.removeData( 'languagefilter' );
+			this.createLanguageFilter();
+
+			this.shouldRecreate = false;
+		},
+
+		/**
+		 * Creates and returns a new debounced version of the passed function,
+		 * which will postpone its execution, until after wait milliseconds have elapsed
+		 * since the last time it was invoked.
+		 *
+		 * @param {Function} fn Function to be debounced.
+		 * @param {Number} wait Wait interval in milliseconds.
+		 * @param {boolean} [immediate] Trigger the function on the leading edge of the wait interval,
+		 * instead of the trailing edge.
+		 * @return {Function} Debounced function.
+		 */
+		debounce: function ( fn, wait, immediate ) {
```

Comment:
this is in practice duplicate of 'delay' method on jquery.uls.languagefilter.js. Since we need this debouncing here and in search input field, how about defining it as $.fn.uls.debounce and reusing in both places?

Reply:
Actually, two functions, `delay()` and `debounce()` are used for different purposes. Debouncing resize event helps greatly with performance by waiting until 300ms of inactivity is over.

Because of how the `keyup` handler is registered and invoked, using `debounce` on keyup event could result in race condition, which can yield incorrect results while searching.

`delay()` just delays the execution of the method, which is executed unconditionally after specified waiting period. I think we should keep both, as methods perform different jobs.

Reply:
The purpose is same, fire the handler of the event after a silence period. debounce is the widely used pattern to debounce search fields. Here also we can use the same.

But in languagefilter,js we combined event handler for tab, escape and character input as single handler. With that, the debounce won't work. We need a non-debounced handler for keyup events for handling control characters and a `change` event handler for actual search query input. This `change` handler should be debounced using the same debounce function we added. in #283, I had started this simplification process several weeks back, not reviewed yet.

Reply:
The change handler that fires search API requests should not use `debounce`: That makes the slightly slow API query even slower by delaying the start of that query by `wait` ms. At most the it can use `throttle` (which I think what `delay` is, but I didn't check) to reduce the amount of requests we start. In-flight requests should be cancelled if possible, but at minimum when they are processed we must check that the query matches the current query, otherwise we have race conditions displaying wrong results when slower API requests complete after fast requests (such as empty search).

Reply:
In the latest commit, one common method is used and that is `debounce`. @santhoshtr was right, the purpose of two methods is same, so I have reorganized the code to have one `debounce` method for both use cases.

Using `delay` is debouncing (not throttling) and is aligned with how language search was working before latest changes to always use API. Canceling seems the right way we should go about subsequent request, but that is not scope of this pull request.

Maybe I should revert the latest changes as the code is likely to be changed soon if we implement canceling and fix race condition.

I was wrong in previous claims on what causes race condition, as the faulty behavior is spotted on `master`. That bug can turn really alarming as searching can be unpredictable and broken at times.